libbpf-rs/src/error.rs

```diff
@@ -16,6 +18,90 @@ pub enum Error {
     Internal(String),
 }
 
+fn errno_to_str(errno: i32) -> Option<&'static str> {
+    let s = match errno {
+        libc::EPERM => stringify!(EPERM),
+        libc::ENOENT => stringify!(ENOENT),
+        libc::ESRCH => stringify!(ESRCH),
+        libc::EINTR => stringify!(EINTR),
+        libc::EIO => stringify!(EIO),
+        libc::ENXIO => stringify!(ENXIO),
+        libc::E2BIG => stringify!(E2BIG),
+        libc::ENOEXEC => stringify!(ENOEXEC),
+        libc::EBADF => stringify!(EBADF),
+        libc::ECHILD => stringify!(ECHILD),
+        libc::EAGAIN => stringify!(EAGAIN),
+        libc::ENOMEM => stringify!(ENOMEM),
+        libc::EACCES => stringify!(EACCES),
+        libc::EFAULT => stringify!(EFAULT),
+        libc::ENOTBLK => stringify!(ENOTBLK),
+        libc::EBUSY => stringify!(EBUSY),
+        libc::EEXIST => stringify!(EEXIST),
+        libc::EXDEV => stringify!(EXDEV),
+        libc::ENODEV => stringify!(ENODEV),
+        libc::ENOTDIR => stringify!(ENOTDIR),
+        libc::EISDIR => stringify!(EISDIR),
+        libc::EINVAL => stringify!(EINVAL),
+        libc::ENFILE => stringify!(ENFILE),
+        libc::EMFILE => stringify!(EMFILE),
+        libc::ENOTTY => stringify!(ENOTTY),
+        libc::ETXTBSY => stringify!(ETXTBSY),
+        libc::EFBIG => stringify!(EFBIG),
+        libc::ENOSPC => stringify!(ENOSPC),
+        libc::ESPIPE => stringify!(ESPIPE),
+        libc::EROFS => stringify!(EROFS),
+        libc::EMLINK => stringify!(EMLINK),
+        libc::EPIPE => stringify!(EPIPE),
+        libc::EDOM => stringify!(EDOM),
+        libc::ERANGE => stringify!(ERANGE),
+        #[allow(unreachable_patterns)]
+        libc::EWOULDBLOCK => stringify!(EWOULDBLOCK),
+        _ => return None,
+    };
+    Some(s)
+}
+
+fn errno_to_text(errno: i32) -> Option<String> {
+    let mut buf = [0 as c_char; 1024];
+    // SAFETY: Our `buf` pointer is always valid because it is coming from a
+    //         Rust array.
+    let rc = unsafe { libc::strerror_r(errno, buf.as_mut_ptr() as *mut _, buf.len() as _) };
+    if rc < 0 {
+        return None;
+    }
+
+    // TODO: Use `CStr::from_bytes_until_nul` once we are at Rust 1.69.
+    let cstr = unsafe { CStr::from_ptr(&buf as _) };
+    Some(String::from_utf8_lossy(cstr.to_bytes()).to_string())
+}
+
+/// Best-effort stringification of errno values to be included in textual
+/// representation of errors.
+fn error_suffix(errno: i32) -> String {
+    match (errno_to_str(errno), errno_to_text(errno)) {
```

Comment:
while errno is always positive, errors returned from lots of libbpf APIs directly are negative. I found it useful in practice to not have to care/remember to negate error codes. So consider stringifying `abs(errno)` to handle both styles of error codes transparently

Reply:
Yeah, we have to make sure that we convert these return values correctly. I am not in favor of just papering over wrong conversions, to be honest.